java/opendcs/src/main/java/org/opendcs/database/DatabaseService.java

```diff
@@ -0,0 +1,115 @@
+    {
+        try
+        {
+            final AuthSource auth = AuthSourceService.getFromString(settings.DbAuthFile);
+            final Properties credentials = auth.getCredentials();
+            return getDatabaseFor(appName, settings, credentials);
+        }
+        catch (AuthException ex)
+        {
+            throw new DatabaseException("Unable to get credentials for database.", ex);
+        }
+    }
+
+    public static OpenDcsDatabase getDatabaseFor(String appName, DecodesSettings settings, Properties credentials) throws DatabaseException
+    {
+        final Iterator<DatabaseProvider> providers = loader.iterator();
+        while (providers.hasNext())
+        {
+            final DatabaseProvider provider = providers.next();
+            if (provider.canCreate(settings))
+            {
+                try
+                {
+                    return provider.createDatabase(appName, settings, credentials);
+                }
+                catch (Exception ex)
+                {
+                    throw new DatabaseException("Unable to create database instance.", ex);
+                }
+            }
+        }
+        throw new DatabaseException("No provider found for database type  " + settings.editDatabaseType);
+    }
+
+    /**
+     * Create database given a valid DataSource. Will load settings from the database.
+     * @param dataSource any valid javax.sql.DataSource. However, a connection pool is expected as multiple connections 
+     *                   may be needed
+     * @return
+     * @throws DatabaseException any error with class initialization or loading of properties
+     */
+    public static OpenDcsDatabase getDatabaseFor(javax.sql.DataSource dataSource) throws DatabaseException
+    {
+        /**
+         * extract properties; requires simple table
+         */
+        DecodesSettings settings = decodesSettingsFromJdbc(dataSource);
+
+        final Iterator<DatabaseProvider> providers = loader.iterator();
+        while (providers.hasNext())
+        {
+            final DatabaseProvider provider = providers.next();
+            if (provider.canCreate(settings))
+            {
+                return provider.createDatabase(dataSource, settings);
+            }
+        }
+        throw new DatabaseException("No provider for database type " + settings.editDatabaseType);
+    }
+
+    /**
+     * Retrieve the decodes settings.
+     *
+     * @param dataSource
+     * @return
+     * @throws DatabaseException
+     */
+    private static DecodesSettings decodesSettingsFromJdbc(javax.sql.DataSource dataSource) throws DatabaseException
+    {
+        DecodesSettings settings = new DecodesSettings();
+        // TODO: Need a special case for the XML database.
+        try (Connection conn = dataSource.getConnection();
+             PreparedStatement stmt = conn.prepareStatement("select name,value from database_properties");
```

Comment:
I don't see any reference to database scripts creating this table

Reply:
Yep, doesn't exist yet. This was a strawman, at least for the moment.

Can either add a specific table though, or create a "database" app in the existing tsdb_properties table.